### DISCUSSION TITLE
Failure proof installation instructions // cannot make this work

### DISCUSSION THREAD

**Discussion by tiagommmelo** [tbabej/taskwiki#447] [Community support] [answered]
Hello. I'm an advanced taskwarrior user, no noob in it (I use timewarrior with it, really comfortable with CLI, tmux and zellij to enhance the experience overall) and I'd like to start using vimwiki/taskwiki in vim or neovim because it would make really easy to set dependent tasks were needed.
However, I just cannot make it work! I searched the web, videos and all of them have different installation instructions... reinstalled several times and still nothing.

### What I tried first?
So, for starters, going step by step on the documentation, I've cloned Vimwiki dev branch:
` git clone https://github.com/vimwiki/vimwiki ~/.vim/bundle/ --branch dev`
Skipped taskwarrior as it is installed (2.6.1) and installed tasklib (without sudo otherwise it throws an error):
`pip3 install tasklib`
Then, cloned taskwiki:
`git clone https://github.com/tools-life/taskwiki ~/.vim/bundle/taskwiki`
After this, when I try to clone vim-plugin-AnsiEsc with the command below:
`git clone https://github.com/powerman/vim-plugin-AnsiEsc ~/.vim/bundle/`
It throws the error ~fatal destination path '/home/<username>/.vim/bundle' already exists and is not an empty directory."
As this is not a requirement, only highly desirable as per what I understood from the documentation, I only tried, then, to use it (even not understanding exactly how I should do it because the documentation is not clear) and it seems to me it is not working or properly installed...

#### Attempt one
I simply did `vim test.md` to create a file; then I put this on the first line:
`==Test proj | project:TestProject==`
I saved and it does nothing (I believe it should have brought tasks from my existing taskwarrior TestProject)

#### Attempt two
I simply opened vim with no extra commands, then issued it the command `:help taskwiki` and it throws me the error "E149: Sorry, no help for taskwiki", so I assume it is not properly installed at this point

#### Attempt three
At this point, I started looking on the web... found a couple of videos showing it in NeoVim, but there are no clear instructions as well... 
Then, I found this useful blogpost (https://sudoedit.com/install-vimwiki-neovim/) and kind of managed to install neovim, vimwiki and taskwiki (I can run now the command `:help taskwiki` and `:help vimwiki` and it actually show me some kind of really nice man page in nvim). It is worth mentioning that my conf file is exact the same as shown in this blog; the only difference is that the "/path/to/my/wiki/" was set to "~/taskwiki" as I have only a "test.md" file there for this purpose. However, when I issue `nvim test.md` (in taskwiki directory), I get this error:

> Error detected while processing BufReadPost Autocommands for "*.md"..function <SNR>9_setup_new_wiki_buffer[19]..vimwiki#u#ft_set[9]..FileType Autocommands for "*"..function <SNR>1_LoadFTPlugin[19]..script /home/tiago_ubuntu/.local/share/nvim/plugged/taskwiki/ftplugin/vimwiki/taskwiki.vim[29]..function provider#python3#Call:
line   18:
Error invoking 'python_execute_file' on channel 3 (python3-script-host):
Traceback (most recent call last):
  File "/home/tiago_ubuntu/.local/share/nvim/plugged/taskwiki/taskwiki/main.py", line 15, in <module>
    from taskwiki import cache as cache_module
  File "/home/tiago_ubuntu/.local/share/nvim/plugged/taskwiki/taskwiki/cache.py", line 5, in <module>
    from taskwiki import preset
  File "/home/tiago_ubuntu/.local/share/nvim/plugged/taskwiki/taskwiki/preset.py", line 5, in <module>
    from taskwiki import util
  File "/home/tiago_ubuntu/.local/share/nvim/plugged/taskwiki/taskwiki/util.py", line 3, in <module>
    from packaging import version
ModuleNotFoundError: No module named 'packaging'
Error invoking 'python_execute' on channel 3 (python3-script-host):
Traceback (most recent call last):
  File "<string>", line 1, in <module>
NameError: name 'WholeBuffer' is not defined
Error invoking 'python_execute' on channel 3 (python3-script-host):
Traceback (most recent call last):
  File "<string>", line 1, in <module>
NameError: name 'cache' is not defined
Press ENTER or type command to continue

So, needless to say that at this point I'm giving up and I'm totally lost... my request is that someone points me to (if it exists) a better installation instruction; it does not matter if the instructions are for neovim or vim... I just would like to actually get it to work... Also, the installation instructions should mention, after the installation, how actually I start taskwiki... if just opening a file is OK (like I tried to do), if it should be another extension other than '.md', if it needs to be in a specific directory, etc...

Thanks so much and if my explanation above gives the impression that I'm frustrated, that's because I am! Thanks again

**Comment by liskin** [comment]
> So, for starters, going step by step on the documentation, I've cloned Vimwiki dev branch:
> ` git clone https://github.com/vimwiki/vimwiki ~/.vim/bundle/ --branch dev`

Hm, looks like taskwiki's readme is really old, written at a time when everyone used https://github.com/tpope/vim-pathogen. That's not the case any more, so you should install vimwiki and taskwiki as you would install any other vim plugin, using whatever vim plugin manager you happen to use. And if you don't use any, then either install vim-pathogen, or use `~/.vim/pack/bundle/start` instead of `~/.vim/bundle`, which is a path where vim's builtin plugin manager expects plugins.


> After this, when I try to clone vim-plugin-AnsiEsc with the command below: `git clone https://github.com/powerman/vim-plugin-AnsiEsc ~/.vim/bundle/` It throws the error ~fatal destination path '/home//.vim/bundle' already exists and is not an empty directory."

This seems like a UX change in git. Perhaps it previously recognised the trailing slash and cloned into `~/.vim/bundle/vim-plugin-AnsiEsc` but it's not documented in the manpage as of now and doesn't work. The following works:
```sh
git -C ~/.vim/bundle/ clone https://github.com/powerman/vim-plugin-AnsiEsc
```

(but do keep in mind that you don't necessarily want to clone into `~/.vim/bundle`)

> I get this error:
> […]
> > ModuleNotFoundError: No module named 'packaging'

This means you need to `pip3 install packaging` as well.

> Also, the installation instructions should mention, after the installation, how actually I start taskwiki... if just opening a file is OK (like I tried to do), if it should be another extension other than '.md', if it needs to be in a specific directory, etc...

The README says taskwiki enhances vimwiki which means you need to setup vimwiki and get familiar with it a bit, but yeah, I suppose the README (and the docs) could use some tender loving care. :-)

**Reply by tiagommmelo** [reply]
Thanks for this answer. I sort of saw the light in the end of the tunnel now.
I've tried a couple of different solutions:

### 1 - Cloning into the other path suggested
I've cloned Vimwiki and AnsiEsc as below:
`git clone https://github.com/vimwiki/vimwiki ~/.vim/pack/bundle/start --branch dev`
`git -C ~/.vim/pack/bundle/start clone https://github.com/powerman/vim-plugin-AnsiEsc`

No issues in cloning, all works. However, when trying to get into vimwiki after that (Leader ww), nothing happens. Also, trying to do a viewport in some other file fails, so either there is a missing step or the plugins cannot be read.

The noob part: I'd need instructions on how to use a vim plugin manager... right now I'm googling for it so I can further check this...

### 2 - Fixing the NeoVim installation
This one lets me see the light in the end of the tunnel :).
I've did as you suggested, which was installing "packaging" `pip3 install packaging`, opened up NeoVim `nvim` and pressed (Leader) ww and voilá! It took me to the empty file "~/taskwiki/index.md". From there, I've typed "* [ ] Taskwiki test task", saved the file and it created the task in Taskwarrior! Getting in and out of visual mode also produced expected results, as per showing and hiding the taskwarriors UID on these tasks.

However, when I try a viewport for an existing tasks, even though it behaves like it should (like hiding the filter in normal mode), when I saved the file to check if the filter would work, I got the following error:

> Error invoking 'python_execute' on channel 3 (python3-script-host):
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/tiago_ubuntu/.local/share/nvim/plugged/taskwiki/taskwiki/errors.py", line 26, in wrapped_function
    original_function(*args, **kwargs)
  File "/home/tiago_ubuntu/.local/share/nvim/plugged/taskwiki/taskwiki/decorators.py", line 10, in wrapped_function
    original_function(*args, **kwargs)
  File "/home/tiago_ubuntu/.local/share/nvim/plugged/taskwiki/taskwiki/main.py", line 64, in update_to_tw
    c.evaluate_viewports()
  File "/home/tiago_ubuntu/.local/share/nvim/plugged/taskwiki/taskwiki/cache.py", line 247, in evaluate_viewports
    port.sync_with_taskwarrior()
  File "/home/tiago_ubuntu/.local/share/nvim/plugged/taskwiki/taskwiki/viewport.py", line 395, in sync_with_taskwarrior
    self.cache.insert_line(str(vimwikitask), added_at)
  File "/home/tiago_ubuntu/.local/share/nvim/plugged/taskwiki/taskwiki/vwtask.py", line 338, in __str__
    ' ' + '!' * self.priority_from_tw_format if self['priority'] else '',
  File "/home/tiago_ubuntu/.local/share/nvim/plugged/taskwiki/taskwiki/vwtask.py", line 244, in priority_from_tw_format
    return convert_priority_from_tw_format(self.task['priority'])
  File "/home/tiago_ubuntu/.local/share/nvim/plugged/taskwiki/taskwiki/vwtask.py", line 15, in convert_priority_from_tw_format
    return {None: None, 'L': 1, 'M': 2, 'H': 3}[priority]
"~/taskwiki/index.md" 3L, 52B written                                                                                                                                                                                                                                        Press ENTER or type command to continue      

Now, if I exclude the viewport, all seems to work fine.

Is there an easy fix for this? Again, I do not mind using vim or neovim, as long as this works. Also, if using VIM, I'd really like to have markdown usage as it seems overall more intuitive

**Reply by liskin** [reply]
`git clone https://github.com/vimwiki/vimwiki ~/.vim/pack/bundle/start --branch dev` — well this clones vimwiki directly into `~/.vim/pack/bundle/start` rather than `~/.vim/pack/bundle/start/vimwiki` where you want it…

**Reply by liskin** [reply, marked as answer]
The other error… are you using custom priorities in taskwarrior by any chance?

(It seems the error output is cut in the middle and half of it is missing so this is really only a wild guess as we don't know what the actual exception is.)

**Reply by tiagommmelo** [reply]
Man, you nailed it.
First, the place where I was cloning the files... once I cloned into the right place, then clearly I could use the plugins into vim as well.
Second, the same error with viewports were occurring in vim as well and I was using custom priorities in Taskwarrior. Once I converted all my tasks (removed all priorities, but added tags with the custom priorities I was using so I do not get lost), then edited ".taskrc" and commented all the lines where custom priorities were set, then there were no more errors in viewports. So all it working now both in Vim and NeoVim.

You rock!!
